<a href="https://colab.research.google.com/github/sarasabino/Accident_Rate_Project/blob/main/Src/Notebooks/00_Data_generation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Data Generation

We initially received the data from the accidents occurred in the company, although it is usefull we will need the data from the staff and the resolution if each employee has had an accident or not. This data will allow us to perform our accident prediction model. 

For that purpose we are going to generate a csv file with random data, but based in the real one, that contains the company's staff by employee. Each row represents an employee, with all its characteristic. In the "Accident" column we will have if that person has had an accident during his time in the company or not. 


In [308]:
! cp /content/drive/MyDrive/'Colab Notebooks'/'TFM Accidentabilidad'/Data/HS_Accidentabilidad.csv .
! cp /content/drive/MyDrive/'Colab Notebooks'/'TFM Accidentabilidad'/Data/'G_Plantas y Tech'.csv .
! cp /content/drive/MyDrive/'Colab Notebooks'/'TFM Accidentabilidad'/Data/Datos_plantilla_fabrica.csv .


We are going to upload the accidents' data and filter by year. We would only work with 2018,2019 and 2020 as the other years are incomplete.

In [309]:
import pandas as pd

accidentes = pd.read_csv('HS_Accidentabilidad.csv', sep=';')

accidentes = accidentes[(accidentes['Año']== 2018) | (accidentes['Año']==2019) | (accidentes['Año']==2020) ]

In [310]:
print("We have a total accidents amount of ",  len(accidentes))

We have a total accidents amount of  4180


In [311]:
accidentes.head() ## we have 104 columns which is a lot of information

ID  -  DATOS DEL SUCESO -  ... Tipo Tema Acta reunión alta baja
2057  2546                    0.0  ...         1                    NaN
2062  2563                    0.0  ...         1                    NaN
2063  2564                    0.0  ...         1                    NaN
2064  2565                    0.0  ...         3                    NaN
2065  2566                    0.0  ...         3                    NaN

[5 rows x 104 columns]

In [312]:
accidentes.columns

Index(['ID', '-  DATOS DEL SUCESO -', 'Codigo accidentado',
       'Fecha introduccion datos', 'Fecha accidente', 'Año',
       'Codigo fecha accidente', 'Planta', 'Código empleado', 'Supervisor',
       ...
       '- PROPUESTA INICIAL DE MEDIDAS A PLANIFICAR -',
       'Medida a planificar 1', 'Medida a planificar 2',
       'Medida a planificar 3', 'Notificacion o investigacion',
       'Modifica notificacion',
       'Se ha dado un suceso similar (en cuanto a sus causas y consecuencias) en el último año en planta?',
       'En caso afirmativo, cual es y como se actuó?', 'Tipo Tema',
       'Acta reunión alta baja'],
      dtype='object', length=104)

In [313]:
#lets check all columns dtypes
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(accidentes.dtypes)

ID                                                                                                     int64
-  DATOS DEL SUCESO -                                                                                float64
Codigo accidentado                                                                                    object
Fecha introduccion datos                                                                              object
Fecha accidente                                                                                       object
Año                                                                                                  float64
Codigo fecha accidente                                                                                 int64
Planta                                                                                                 int64
Código empleado                                                                                       object
Supervisor         

In [314]:
accidentes['Gravedad potencial'].unique()

array(['Muy grave', 'Leve', 'Grave', nan], dtype=object)

In [315]:
## we are going to filter to get the most important columns
staff = accidentes[['Codigo accidentado', 'Puesto habitual de trabajo','Turno','Edad','Género','Planta', 'Tipo accidente', 'Tipo trabajador', 'Gravedad potencial', 'Lugar o seccion accidente']]
staff.head()

Codigo accidentado  ...      Lugar o seccion accidente
2057       HS-CLMB_2546  ...              Células Soldadura
2062       HS-CEPA_2563  ...                          Tubos
2063       HS-GNVT_2564  ...                          Tubos
2064       HS-GNVT_2565  ...  Almacén de Producto Terminado
2065       HS-DNGG_2566  ...                       Blanking

[5 rows x 10 columns]

Now we are going to upload the Plants and Tech file, which contains the relation between the fabric the tecnology and the country. We can join this table with the staff table by Planta - ID

In [316]:
planta = pd.read_csv('G_Plantas y Tech.csv', sep=';')
planta.drop(columns={'Activo'}, inplace=True)
planta.head()

ID    Fabrica          Tecnología       Pais
0   1  Argentina            Blanking  Argentina
1   2  Argentina  Corte Longitudinal  Argentina
2   3  Argentina   Corte Transversal  Argentina
3   4    Arizona   Células Soldadura       EEUU
4   5    Arizona         Corte Láser       EEUU

In [317]:
# we merge both df so we have all information in staff df with a left join in case we dont have plants for some rows

staff = staff.merge(planta, left_on='Planta', right_on='ID', how='left')


In [318]:
staff.drop(columns={'ID'}, inplace=True)

staff.head()

Codigo accidentado Puesto habitual de trabajo  ...   Tecnología    Pais
0       HS-CLMB_2546    Operario de Puente Grúa  ...     Skinpass  España
1       HS-CEPA_2563                  Engraving  ...     Recocido  España
2       HS-GNVT_2564                Titular L-8  ...     Skinpass  España
3       HS-GNVT_2565               Carretillero  ...     Skinpass  España
4       HS-DNGG_2566                        NaN  ...  Corte Láser  España

[5 rows x 13 columns]

In [319]:
len(staff) ## we have same amount as before

4180

Now we are going to upload some staff data to get the number of trainning hours per plant and have a reference of the quantity of employees we want to generate. 

This data is aggregated so it is not valid for our model.

In [320]:
rrhh_plant = pd.read_csv('Datos_plantilla_fabrica.csv', sep=';')

rrhh_plant['Año'].unique()

array([2017., 2016., 2018., 2019.,   nan, 2020., 2021.])

In [321]:
# we are getting only 2020 for our analysis to have a reference

rrhh_plant = rrhh_plant[rrhh_plant['Año']==2020]


In [322]:
rrhh_plant.head()

ID  ... Empleados Propios - duración determinada
2615  3819  ...                                     14.0
2616  3820  ...                                     38.0
2617  3821  ...                                      3.0
2618  3822  ...                                      2.0
2619  3823  ...                                     28.0

[5 rows x 16 columns]

In [323]:
a = 545613.01
round(a)

545613

In [324]:
lista = ['Horas Presencia Efectiva Propios', 'Horas Presencia Efectiva Subcontratados', 'Horas Presencia Efectiva ETTs','Horas Formacion Seguridad Propios' ,'Horas Formacion Seguridad ETTs', 'Horas Formacion General Propios'  ]

for column in lista:  
  rrhh_plant[column] = rrhh_plant[column].apply(lambda x: x.replace(',', '.')).astype(float)
  rrhh_plant[column] = rrhh_plant[column].apply(lambda x: round(x))

In [325]:
rrhh_plant.dtypes

ID                                            int64
Planta                                       object
Año                                         float64
Mes                                          object
Empleados Propios                           float64
Empleados Ajenos                            float64
Empleados ETTS                              float64
Empleados Subcontrata                       float64
Horas Presencia Efectiva Propios              int64
Horas Presencia Efectiva Subcontratados       int64
Horas Presencia Efectiva ETTs                 int64
Horas Formacion Seguridad Propios             int64
Horas Formacion Seguridad ETTs                int64
Horas Formacion General Propios               int64
Empleados Propios - indefinido              float64
Empleados Propios - duración determinada    float64
dtype: object

In [326]:
trainning_hours = rrhh_plant[['Planta','Horas Presencia Efectiva Subcontratados', 'Horas Presencia Efectiva ETTs', 'Horas Formacion Seguridad Propios', 'Horas Formacion Seguridad ETTs']]

In [327]:
trainning_hours.rename(columns={'Planta':'Fabrica'}, inplace=True)
trainning_hours.head(1)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Fabrica  ...  Horas Formacion Seguridad ETTs
2615  Barcelona  ...                              23

[1 rows x 5 columns]

In [328]:
trainning_hours = trainning_hours.groupby('Fabrica').agg('sum').reset_index() ## we get the nº of trainning hours by fabric
trainning_hours.head()

Fabrica  ...  Horas Formacion Seguridad ETTs
0          Addimen  ...                               0
1        Argentina  ...                               0
2          Arizona  ...                               0
3  Baja California  ...                               0
4        Barcelona  ...                             152

[5 rows x 5 columns]

In [329]:
staff = staff.merge(trainning_hours, on='Fabrica', how='left') # adding the training hours to staff df
staff.head()

Codigo accidentado  ... Horas Formacion Seguridad ETTs
0       HS-CLMB_2546  ...                          829.0
1       HS-CEPA_2563  ...                          829.0
2       HS-GNVT_2564  ...                          152.0
3       HS-GNVT_2565  ...                          152.0
4       HS-DNGG_2566  ...                            NaN

[5 rows x 17 columns]

Now we are going to calculate the proportion in 2020 between people and accidents occur

In [330]:
employees = pd.read_csv('Datos_plantilla_fabrica.csv', sep=';')
employees = employees[['Año', 'Empleados Propios', 'Empleados Ajenos', 'Empleados ETTS',  'Empleados Subcontrata']]
employees = employees.fillna(0)
employees = employees.groupby(['Año']).agg('sum')
employees.head()

Empleados Propios  ...  Empleados Subcontrata
Año                        ...                       
0.0               15754.0  ...                  786.0
2016.0            40609.0  ...                    0.0
2017.0            43291.0  ...                    0.0
2018.0            49865.0  ...                 2993.0
2019.0            50625.0  ...                 2697.0

[5 rows x 4 columns]

In [331]:
accidentes = pd.read_csv('HS_Accidentabilidad.csv', sep=';')
accidentes = accidentes[['Año', 'Codigo accidentado']]
accidentes = accidentes.groupby('Año').agg('count').reset_index()
accidentes.rename(columns={'Codigo accidentado':'Accidentes'}, inplace=True)
accidentes.head()

Año  Accidentes
0  2015.0         202
1  2016.0        1335
2  2017.0        1203
3  2018.0        1532
4  2019.0        1526

In [332]:
employees = employees.merge(accidentes, on='Año', how='inner')
employees.head()

Año  Empleados Propios  ...  Empleados Subcontrata  Accidentes
0  2016.0            40609.0  ...                    0.0        1335
1  2017.0            43291.0  ...                    0.0        1203
2  2018.0            49865.0  ...                 2993.0        1532
3  2019.0            50625.0  ...                 2697.0        1526
4  2020.0            65644.0  ...                 1335.0        1122

[5 rows x 6 columns]

In [333]:
employees.dtypes

Año                      float64
Empleados Propios        float64
Empleados Ajenos         float64
Empleados ETTS           float64
Empleados Subcontrata    float64
Accidentes                 int64
dtype: object

In [334]:
employees['Total_emp'] = employees['Empleados Propios'] + employees['Empleados Ajenos'] + employees['Empleados ETTS'] + employees['Empleados Subcontrata']

In [335]:
employees = employees[['Año', 'Accidentes', 'Total_emp']]

employees['perc_acc'] = employees['Accidentes'] / employees['Total_emp'] * 100
employees.head()

Año  Accidentes  Total_emp  perc_acc
0  2016.0        1335    49769.0  2.682393
1  2017.0        1203    53085.0  2.266177
2  2018.0        1532    71795.0  2.133853
3  2019.0        1526    71672.0  2.129144
4  2020.0        1122    79866.0  1.404853

For our analysis we are going to suppose we have a year staff, with all people. 
We would have 79900 people with around 7000 people who have suffered an accident on the past years.

In [336]:
len(staff)

4180

In [337]:
#we add a new column to indetify those people who have suffered an accident in our dataset
staff['Accident'] = 'Yes'
staff.head()

Codigo accidentado  ... Accident
0       HS-CLMB_2546  ...      Yes
1       HS-CEPA_2563  ...      Yes
2       HS-GNVT_2564  ...      Yes
3       HS-GNVT_2565  ...      Yes
4       HS-DNGG_2566  ...      Yes

[5 rows x 18 columns]

In [338]:
#plant data we dont have any null values
staff['Planta'].unique()

array([20, 19, 12, 41,  7,  1, 13, 28,  2,  5, 26, 16, 25, 15, 36, 35, 33,
       53, 54,  4, 14, 34, 27, 29, 32, 57,  6, 56, 30, 55, 46, 43, 45, 18,
       44, 63, 47, 84])

In [339]:
##work position

positions = staff['Puesto habitual de trabajo'].unique()
len(positions)

794

In [340]:
import random
import numpy as np 

staff['wk_position'] = np.random.choice(positions, size=len(staff))

In [341]:
staff['wk_position'].unique()

array(['Gruista LONGI004 y LONGI008', 'Operador Puente-Grúa', 'hanger',
       'Titular Circulares', 'Control Calidad', 'Operador Logistico',
       'Embaladora 2', 'AYUDANTE PRENS001', 'Gruista perfiladoras',
       'packaking -road safety', 'galvanizado',
       'operator on packaging line', 'GRUISTA - EMBALADOR',
       'Operadores limpieza de caldera y gruista de caldera',
       'Operador de Produção II', 'Titular Línea Corte Prensa II',
       'Inspector de Calidad', 'empaquetador perfiladora de tubos',
       'EMBALAJE MANUAL', 'Kranführer H1 Nord',
       'Staplerfahrer Halle 2 und 3', 'Recibidor línea 3mts',
       'Auxiliar TRANS003', 'Perfiles abiertos', 'Operario de Linea',
       'Utillero', 'Operario de mantenimiento', 'welding for polygon',
       'Operador II', 'LCP 03', 'Movimentação',
       'Material  Handler / Slitting Operator', 'OPERARIO DE CARGA',
       'Almacenes', 'Auxiliar de Producción',
       'Mitarbeiter Qualitätsmanagement', 'Flejador/ ayudante TRANS007'

In [342]:
staff.drop(columns={'Puesto habitual de trabajo'}, inplace=True)

In [343]:
#gender

staff.rename(columns={'Género':'Gender'}, inplace=True)


In [344]:
staff['Gender'].unique()

array(['Hombre', nan, 'Mujer'], dtype=object)

In [345]:
staff['Gender'] = staff['Gender'].fillna('Hombre')

In [346]:
len(staff[staff['Gender'].isnull()])

0

In [347]:
staff['Gender'].unique()

array(['Hombre', 'Mujer'], dtype=object)

Checking if we have null values in turno

In [348]:
staff.Turno.isnull().any()

False

Checking for nulls in age:

As we have null values we are goign to reassing the age range based on the proportion of the tables we hav analyzed

In [349]:
staff.Edad.isnull().any()

True

In [350]:
len(staff[staff.Edad.isnull()])

1855

In [351]:
rangos_edad = ['Menor 30', 'Entre 30-50', 'Mayor 50']
weights = {'Hombre':[0.2, 0.44, 0.36], 'Mujer':[0.2,0.55,0.25]}

In [352]:
np.random.seed(0)

def apply_randoms(x):
  key = x['Gender'].iat[0]
  return pd.Series(np.random.choice(rangos_edad, size=len(x), p=weights[key]))

In [353]:
staff['Age'] = staff.groupby('Gender').apply(apply_randoms)\
       .rename('Age').reset_index().drop('level_1', 1)['Age']

In [354]:
staff.drop(columns={'Edad'}, inplace=True)

Checking is there is any row with null at type of accident

In [355]:
staff['Tipo accidente'].isnull().any()

False

Worker type have null values at some rows

In [356]:
staff['Tipo trabajador'].isnull().any()

True

In [357]:
staff['Tipo trabajador'].unique()

array(['ETT', 'Indefinido', 'Duración Determinada',
       'Incidente (no aplica)', 'Subcontratado', 'Prácticas', nan,
       'Cedidos de otras plantas', 'No aplica', 'Indefinido con preaviso'],
      dtype=object)

In [358]:
staff[staff['Tipo trabajador'].isnull()] #As we only have two values null we are going to assign directly 'Indefinido'

Codigo accidentado            Turno  ...         wk_position       Age
2831       HS-NVRR_8597  Turno de mañana  ...                  /   Menor 30
2832       HS-NVRR_8598  Turno de mañana  ...  Packaging Operator  Mayor 50

[2 rows x 18 columns]

In [359]:
staff['Tipo trabajador'] = staff['Tipo trabajador'].fillna('Indefinido')

In [360]:
staff['Tipo trabajador'].isnull().any()

False

Potential Severity columns has null values as well

In [361]:
staff[staff['Accident']=='Yes']['Gravedad potencial'].isnull().any()

True

In [362]:
len(staff[(staff['Gravedad potencial'].isnull()) & (staff['Accident']=='Yes')])

1009

In [363]:
## we are going to assign it randomly
staff['Gravedad potencial'].unique()
severity = ['Muy grave', 'Leve', 'Grave']

In [364]:
for i in staff['Gravedad potencial']:

  if str(i) == 'nan':
    staff['Severity'] = np.random.choice(severity, size=len(staff))
  else:
    staff['Severity'] = i


In [365]:
staff['Severity'].isnull().any()

False

In [366]:
staff['Severity'].unique()

array(['Leve', 'Muy grave', 'Grave'], dtype=object)

In [367]:
staff.drop(columns={'Gravedad potencial'}, inplace=True)

In [368]:
staff['Severity'].isnull().any()

False

Aciccent's place/section

In [369]:
staff['Lugar o seccion accidente'].isnull().any()

True

In [370]:
len(staff[staff['Lugar o seccion accidente'].isnull()])

63

In [371]:
staff['Lugar o seccion accidente'].unique()

array(['Células Soldadura', 'Tubos', 'Almacén de Producto Terminado',
       'Blanking', 'Corte Longitudinal', 'Corte Transversal',
       'Almacén de Materia Prima', 'Cizalla', 'Decapado', 'Puente Grúa',
       'Carpinteria', 'Galvanizado Inmersión', nan, 'Maquinaria Auxiliar',
       'Corte Plasma', 'Perfiles abiertos', 'Corte Láser',
       'Almacen productos quimicos', 'Corte Láser 3D',
       'Instalaciones exteriores y auxiliares', 'Mantenimiento', 'Prensa',
       'Almacén producto semielaborado o en producción', 'Carretilla',
       'Sierra', 'Oficinas administrativas', 'Área de embalado',
       'Punzonado', 'Soldadura Manual', 'Limpieza/Prepintado',
       'Galvanizado continuo', 'Perforado', 'Coil to Coil',
       'Tailor Welding', 'Galvanizado ', 'Almacen de recambios',
       'Galvanizado', 'Electrozincado', 'Pintura', 'General / Auxiliar',
       'Perfilado de barrera y postes', 'Skinpass', 'Slitting ',
       'General', 'Press Blanking', 'Laminado', 'Ablation Line',
    

In [372]:
place_acc = ['Células Soldadura', 'Tubos', 'Almacén de Producto Terminado',
       'Blanking', 'Corte Longitudinal', 'Corte Transversal',
       'Almacén de Materia Prima', 'Cizalla', 'Decapado', 'Puente Grúa',
       'Carpinteria', 'Galvanizado Inmersión',  'Maquinaria Auxiliar',
       'Corte Plasma', 'Perfiles abiertos', 'Corte Láser',
       'Almacen productos quimicos', 'Corte Láser 3D',
       'Instalaciones exteriores y auxiliares', 'Mantenimiento', 'Prensa',
       'Almacén producto semielaborado o en producción', 'Carretilla',
       'Sierra', 'Oficinas administrativas', 'Área de embalado',
       'Punzonado', 'Soldadura Manual', 'Limpieza/Prepintado',
       'Galvanizado continuo', 'Perforado', 'Coil to Coil',
       'Tailor Welding', 'Galvanizado ', 'Almacen de recambios',
       'Galvanizado', 'Electrozincado', 'Pintura', 'General / Auxiliar',
       'Perfilado de barrera y postes', 'Skinpass', 'Slitting ',
       'General', 'Press Blanking', 'Laminado', 'Ablation Line',
       'Slitting', 'D-Ring Line', 'Multi Strand']

In [373]:
for i in staff['Lugar o seccion accidente']:

  if str(i) == 'nan':
    staff['Accident_place'] = np.random.choice(place_acc, size=len(staff))
  else:
    staff['Accident_place'] = i

In [374]:
staff['Accident_place'].isnull().any()

False

In [375]:
staff.drop(columns={'Lugar o seccion accidente'}, inplace=True)

There arent any null values in Fabric column, technology and country fields are from fabric so the will be filled as well

In [376]:
staff['Fabrica'].isnull().any()

False

Now we are going to check the columns of efective presence hours

In [377]:
staff['Horas Presencia Efectiva Subcontratados'].isnull().any()

True

In [378]:
staff['Horas Presencia Efectiva Subcontratados'] = staff['Horas Presencia Efectiva Subcontratados'].fillna(0)

In [379]:
staff['Horas Presencia Efectiva ETTs'].isnull().any()

True

In [380]:
staff['Horas Presencia Efectiva ETTs'] = staff['Horas Presencia Efectiva ETTs'].fillna(0)

In [381]:
staff['Horas Formacion Seguridad Propios'].isnull().any()

True

In [382]:
staff['Horas Formacion Seguridad Propios'] = staff['Horas Formacion Seguridad Propios'].fillna(0)

In [383]:
staff['Horas Formacion Seguridad ETTs'].isnull().any()

True

In [384]:
staff['Horas Formacion Seguridad ETTs'] = staff['Horas Formacion Seguridad ETTs'].fillna(0)

In [385]:
staff.head()

Codigo accidentado            Turno  ...   Severity  Accident_place
0       HS-CLMB_2546  Turno de mañana  ...       Leve   Mantenimiento
1       HS-CEPA_2563  Turno de mañana  ...  Muy grave   Mantenimiento
2       HS-GNVT_2564   Turno de tarde  ...  Muy grave   Mantenimiento
3       HS-GNVT_2565  Turno de mañana  ...      Grave   Mantenimiento
4       HS-DNGG_2566  Turno de mañana  ...      Grave   Mantenimiento

[5 rows x 18 columns]

Now we are going to generate the other part of our staff dataframe which contains those people who havent had an accident

In [386]:
len(staff)

4180

In [387]:
employees = rrhh_plant[['Año', 'Mes', 'Empleados Propios', 'Empleados Ajenos','Empleados ETTS','Empleados Subcontrata']]
employees  = employees[employees['Mes']=='ENE']

employees['Total'] = employees['Empleados Propios'] + employees['Empleados Ajenos'] + employees['Empleados ETTS'] + employees['Empleados Subcontrata']

employees = employees.groupby(['Año', 'Mes']).agg('sum')
employees['Total']

Año     Mes
2020.0  ENE    7070.0
Name: Total, dtype: float64

So the number of employees in 2020 7070 and we are going to take that as reference. 

As we are taking 3 years of accidents we are going to have the same proportion on the staff to be able to evaluate it. So our total number of employees in the staff table will be 21.210, 4180 accidents. 

We are going to generate now the 17030 employees who didnt have any accident

In [388]:
staff_without = pd.DataFrame(columns=staff.columns)

staff_without

Empty DataFrame
Columns: [Codigo accidentado, Turno, Gender, Planta, Tipo accidente, Tipo trabajador, Fabrica, Tecnología, Pais, Horas Presencia Efectiva Subcontratados, Horas Presencia Efectiva ETTs, Horas Formacion Seguridad Propios, Horas Formacion Seguridad ETTs, Accident, wk_position, Age, Severity, Accident_place]
Index: []

In [389]:
def create_rndm(values, column):
  staff_without[column] = np.random.choice(values, size=17030)

In [390]:
turno  = ['Turno de mañana', 'Turno de tarde', 'Turno de noche']

create_rndm(turno, 'Turno')

To create the gender columns we are going to apply weights based on the data we had from staff

In [391]:
gender = ['Hombre', 'Mujer']
weights = [0.90,0.1]

np.random.seed(0)

staff_without['Gender'] = random.choices(gender, weights=weights, k=len(staff_without))

In [392]:
planta_list = staff.Planta.unique()
create_rndm(planta_list, 'Planta')

In [393]:
staff_without.head(1)

Codigo accidentado           Turno  Gender  ...  Age Severity Accident_place
0                NaN  Turno de tarde  Hombre  ...  NaN      NaN            NaN

[1 rows x 18 columns]

In [394]:
staff_without.drop(columns={'Fabrica', 'Tecnología', 'Pais'}, inplace=True)

In [395]:
staff_without = staff_without.merge(planta, left_on='Planta', right_on='ID', how='left')
staff_without.head(2)

Codigo accidentado            Turno  Gender  ...   Fabrica   Tecnología    Pais
0                NaN   Turno de tarde  Hombre  ...    Burgos     Skinpass  España
1                NaN  Turno de mañana  Hombre  ...  Asturias  Corte Láser  España

[2 rows x 19 columns]

In [396]:
len(staff_without)

17030

In [397]:
tipo_trabj = staff['Tipo trabajador'].unique()
create_rndm(tipo_trabj, 'Tipo trabajador')

In [398]:
staff_without.drop(columns=['Horas Presencia Efectiva Subcontratados','Horas Presencia Efectiva ETTs', 'Horas Formacion Seguridad Propios', 'Horas Formacion Seguridad ETTs' ], inplace=True)

In [399]:
staff_without = staff_without.merge(trainning_hours, on='Fabrica', how='left')

In [400]:
staff_without.drop(columns={'ID'}, inplace=True)

In [401]:
staff_without['Accident'] = 'No'

In [402]:
pos_wk = staff['wk_position'].unique()

create_rndm(pos_wk, 'wk_position')

In [403]:
rangos_edad = ['Menor 30', 'Entre 30-50', 'Mayor 50']
weights = {'Hombre':[0.2, 0.44, 0.36], 'Mujer':[0.2,0.55,0.25]}

np.random.seed(0)

staff_without['Age'] = staff_without.groupby('Gender').apply(apply_randoms)\
       .rename('Age').reset_index().drop('level_1', 1)['Age']

In [404]:
place = staff['Accident_place'].unique()

create_rndm(place, 'Accident_place')

In [405]:
staff_without.head(2) ##cod accidentado is the only column we are going to leave blank as these employees doestn have any accident

Codigo accidentado  ... Horas Formacion Seguridad ETTs
0                NaN  ...                          829.0
1                NaN  ...                            NaN

[2 rows x 18 columns]

Now we are going to concat both dataframes on a resulting dataframe with length of 21.210

In [406]:
staff = pd.concat([staff, staff_without])
len(staff)

21210

In [407]:
staff.iloc[0]

Codigo accidentado                                        HS-CLMB_2546
Turno                                                  Turno de mañana
Gender                                                          Hombre
Planta                                                              20
Tipo accidente                                           Accidente c/b
Tipo trabajador                                                    ETT
Fabrica                                                         Burgos
Tecnología                                                    Skinpass
Pais                                                            España
Horas Presencia Efectiva Subcontratados                          25249
Horas Presencia Efectiva ETTs                                    42485
Horas Formacion Seguridad Propios                                 1419
Horas Formacion Seguridad ETTs                                     829
Accident                                                           Yes
wk_pos

In [408]:
from google.colab import files

staff.to_csv('Total_staff_by_employee.csv', sep=';')
files.download("Total_staff_by_employee.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>